notes.md

```diff
@@ -0,0 +1,1909 @@
+
+If you have an attribute that you don't want to be serialized you should not store it as an attribute on tensor or Parameter but instead it is recommended to use `torch.utils.weak.WeakTensorKeyDictionary`
+
+```Python
+>>> foo_dict = weak.WeakTensorKeyDictionary()
+>>> foo_dict[a] = 'hey'
+>>> print(foo_dict[a])
+hey
+```
+
+### **Algorithms `{Adadelta, Adagrad, Adam, Adamax, AdamW, ASGD, NAdam, RAdam, RMSProp, RProp, SGD}` default to faster `foreach` implementation when on CUDA + differentiable=`False`**
+
+When applicable, this changes the default behavior of `step()` and anything that calls into `adadelta(...)`, `adagrad(...)`, `adam(...)`, `adamax(...)`, `adamw(...)`, `asgd(...)`, `nadam(...)`, `radam(...)`, `rmsprop(...)`, `rprop(...)`, `sgd(...)` directly to use the `foreach` implementation instead of the for-loop for better performance. Applicable means
+
+1. The user has not specified kwargs relating to implementation (`foreach`, `fused`, or `differentiable`),
+2. All tensors are native tensors (not subclasses) and on CUDA,
+3. `torch.jit.is_scripting` is `False`.
+
+When these conditions are satisfied, the implementation used will match the implementation used when one passes `foreach=True`. The user defined flag for `foreach` will NOT be overwritten in order to preserve user selections. For more details, check the [documentation](https://pytorch.org/docs/stable/optim.html#algorithms). There should be no significant differences between the results returned by these optimizers. To revert to the old behavior, say, for `adam`, pass in `adam(..., foreach=False, ...)` or initialize `Adam` with `Adam(..., foreach=False, ...)`.
+
+Pull Requests: #92306, #92716, #92723,#92724, #92726, #92727, #92728, #92715, #91896, #92730, #90865, #93184, #92181, #92923, #95415, #95818, #95811
+
+### **`torch.nn.utils.stateless.functional_call` now respects tied weights (#90477)**
+
+Assume a module has two tied weights, x and x_tied. Previously, invoking `functional_call(module, parameters_and_buffers, args, kwargs=None, *, strict=False)` with a parameter dictionary of only one of the tied weights would result in the other one(s) not being updated.
+
+We’ve changed the behavior so that providing one of the tied weights in the parameter dictionary will update all other tied weights. If you would like the behavior in previous versions of PyTorch, please set `tie_weights=False`.
+
+Please also see the related deprecation section "torch.nn.stateless.functional_call in favor of torch.func.functional_call".
+
+<table>
+<tr>
+<th>1.13</th>
+<th>2.0</th>
+</tr>
+<tr>
+<td>
+
+```Python
+>>> class Foo(nn.Module):
+...    def __init__(self):
+...        super().__init__()
+...        self.x = nn.Parameter(torch.zeros([]))
+...        self.x_tied = self.x
+...
+...    def forward(self, inp):
+...        return self.x + self.x_tied
+
+>>> foo = Foo()
+>>> params = {'x': torch.ones([])}
+>>> result = torch.nn.utils.stateless.functional_call(foo, params, torch.randn([]))
```

Comment:
make these lines less long so that the side by side code looks better

Reply:
This is true in a few of these, you can manually add breaklines to be sure everything fits nicely in terms of width.